### PR TITLE
1.x: concat reduce overhead when streaming a source

### PR DESCRIPTION
**Pull request by akarnokd**
This PR reduces the request tracking overhead of `concat` by tracking the produced item count in a plain field and subtracting it from the arbiter and requested values only before the inner source completes. So instead of N decrementAndGet call, we have 1 addAndGet(-N) per source. 

I've added a perf class to measure the difference.

(Intel Celeron 1005M @ 2GHz, Windows 7 x64, Java 8u66)

image

The throughput increased considerably, although I would have expected more, especially in the 1M case where the subscription overhead doesn't matter.

I'll do further investigation on it and post a follow-up PR if this gets merged in the meantime.

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
:+1:

File: src/main/java/rx/internal/operators/OperatorConcat.java [hunk 1]
```diff
@@ -179,6 +187,14 @@ void subscribeNext() {
                 }
             }
         }
+        
+        void produced(long c) {
+            arbiter.produced(c);
+            AtomicLong requestedField = requested;
+            if (requestedField.get() != Long.MAX_VALUE) {
+                requestedField.addAndGet(-c);
```
Comment (zsxwing) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: Call `getAndAddRequest` here? In case `requested` is changed to `Long.MAX_VALUE` at the same time.

Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: In normal use, such change never happens; consumers either start with MAX_VALUE or some prefetch value, plus you have to be very precise to hit the narrow window between the two lines.

Comment (davidmoten) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: I'd prefer to see correctness. Speculating on the probability of a race
condition seems like dangerous territory to be entering.

Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: This line is correct as it is, what we are talking about is considering the case where one wants to subvert the optimization by racing with get().

Comment (davidmoten) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: I've a very narrow focus on this method so I could well be missing something but normally once `requested` is set to Long.MAX_VALUE we don't reduce its value and @zsxwing has indicated this possibility. I don't think it should be best endeavours as in I would expect to see a CAS loop here. Is this a bit fussy? Do we do this in lots of places already? (I thought a lot had been cleaned up).

Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: The keyword is loop which can lenghten due to failed CAS in bounded mode or safepoint kicking in. AddAndGet is a one step atomic operation.

Comment (zsxwing) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: So no place will update `requested` at the same time. Right?

**Comment by akarnokd** [comment]
There is a non-zero chance there, but it is very small. If it happens, all the effect it will have is to make the operator decrement the requested amount from then on, which adds ~40 cycle overhead after each source completes.

Updated the PR anyway to do a CAS loop on production to remove all worries about the case.

**Comment by zsxwing** [comment]
:+1:

**Comment by akarnokd** [comment]
@zsxwing, @davidmoten are you satisfied with the name `BackpressureUtils.produced`?

File: src/main/java/rx/internal/operators/OperatorConcat.java [hunk 2]
```diff
@@ -112,9 +112,19 @@ public void onStart() {
...
+            if (n <= 0) return;
             // we track 'requested' so we know whether we should subscribe the next or not
-            long previous = BackpressureUtils.getAndAddRequest(requested, n);
+            
+            final AtomicLong requestedField = requested;
+            
+            long previous;
+            
+            if (requestedField.get() != Long.MAX_VALUE) {
+                previous = BackpressureUtils.getAndAddRequest(requestedField, n);
```
Comment (davidmoten) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: This path involves two `AtomicLong.get()`. Can't the whole block just be replaced by 
`long previous = BackpressureUtils.getAndAddRequest(requestedField, n);`?

Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: No, the `getAndAddRequest` doesn't check for max-value and always does the CAS. I'd leave this as is and have a separate PR where the `getAndAddRequest` is changed and all use places are verified/fixed.

File: src/main/java/rx/internal/operators/OperatorConcat.java [hunk 3]
```diff
@@ -187,6 +200,8 @@ void subscribeNext() {
         private final ConcatSubscriber<T> parent;
         private final AtomicInteger once = new AtomicInteger();
```
Comment (davidmoten) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: Why not `AtomicBoolean`?

File: src/main/java/rx/internal/operators/OperatorConcat.java [hunk 4]
```diff
@@ -212,6 +227,11 @@ public void onError(Throwable e) {
         @Override
         public void onCompleted() {
             if (once.compareAndSet(0, 1)) {
+                // signal the production count at once instead of one by one
+                long c = produced;
+                if (c != 0L) {
```
Comment (davidmoten) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: Move the `c!=0` test into the `produced` method? Don't see why you need to assign `produced` to `c` as it can't be changed by another thread due Observable contract.

Comment (stevegury) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: I think @davidmoten is right here, I don't see why you need to capture `produced`.

File: src/main/java/rx/internal/operators/OperatorConcat.java (hunk 3 above)
Comment (stevegury) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: AtomicBoolean is very similar to AtomicInteger since it uses under the hood an integer.
But I agree that AtomicBoolean makes the intent of the code more explicit.

Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: internally, `AtomicBoolean` is implemented the same as AtomicInteger, but sure, can be changed to `AtomicBoolean`, by a separate PR.

**Comment by stevegury** [comment]
Just a few comments, otherwise :+1:.

File: src/main/java/rx/internal/operators/OperatorConcat.java (hunk 4 above)
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorConcat.java]: Because of the compareAndSet, the read of parent field would happen all the time, including empty sources. This also affects produced which may or may not be register-cached.

**Comment by akarnokd** [comment]
Changed to AtomicBoolean and moved c != 0 into parent because the parent was used for completing the inner anyway, I overlooked it at first.

**Comment by stevegury** [comment]
:+1:
